### PR TITLE
[Core] Disallow callable for job submission api

### LINKED ISSUE #41493: [core][runtimenv] worker_process_setup_hook not working with ray job
### What happened + What you expected to happen

and maybe not working with ray serve config as well. 

```
(base) ray@ip-10-0-54-240:~/default$ python run.py 
Traceback (most recent call last):
  File "/home/ray/default/run.py", line 15, in <module>
    job = client.submit_job(entrypoint="python script.py", runtime_env={"worker_process_setup_hook": lambda: logging_setup})
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/dashboard/modules/job/sdk.py", line 238, in submit_job
    r = self._do_request("POST", "/api/jobs/", json_data=json_data)
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/dashboard/modules/dashboard_sdk.py", line 303, in _do_request
    return requests.request(
  ....
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type function is not JSON serializable
```

### Versions / Dependencies

master

### Reproduction script

```
import ray
import logging

def logging_setup():
   logger = logging.getLogger("ray")
   logger.setLevel(logging.DEBUG)

job = client.submit_job(entrypoint="python script.py", runtime_env={"worker_process_setup_hook": lambda: logging_setup})
```

### Issue Severity

None

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

Passing callable type to job submission API should be an anti pattern because you need to serialize the callable from the job submission side, and the cluster side can easily have a different environment when you deserialize it. 

The best practice is we only allow string (module name) for worker_process_setup_hook when it is passed from job submission code. Note that worker_process_setup_hook is allowed to be overwritten (without affecting other runtime env) if it is specified via ray.init. 

## Related issue number

Closes https://github.com/ray-project/ray/issues/41493

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rickyyx** [comment]
cc @architkulkarni ?

File: python/ray/runtime_env/runtime_env.py [hunk 1]
```diff
@@ -237,9 +237,11 @@ class MyClass:
         env_vars: Environment variables to set.
         worker_process_setup_hook: (Experimental) The setup hook that's
             called after workers start and before Tasks and Actors are scheduled.
-            The value has to be a callable when passed to the Job, Task, or Actor.
-            The callable is then exported and this value is converted to
-            the setup hook's function name for observability.
+            A module name (string type) or callable (function) can be passed.
+            When a module name is passed, Ray worker should be able to access the
+            module name. When a callable is passed, callable should be serializable.
+            When a runtime env is specified by job submission API,
```
Comment (rickyyx) [inline python/ray/runtime_env/runtime_env.py:243]: is this something we could callout better? This feels rather tricky.

**Review by rickyyx — APPROVED** [review]

File: dashboard/modules/job/sdk.py [hunk 2]
```diff
@@ -214,6 +214,17 @@ def submit_job(
...
 
+        # Verify worker_process_setup_hook type.
+        setup_hook = runtime_env.get("worker_process_setup_hook")
+        if setup_hook and not isinstance(setup_hook, str):
+            raise ValueError(
+                f"Invalid type {type(setup_hook)} for `worker_process_setup_hook`. "
+                "When a job submission API is used, `worker_process_setup_hook` "
+                "only allows a string type (module name). "
+                "Specify `worker_process_setup_hook` via "
+                "ray.init within a driver to use a function type. "
```
Comment (rickyyx) [inline dashboard/modules/job/sdk.py:225]: `Callable`
```suggestion
                "ray.init within a driver to use a Callable type. "
```

**Review by architkulkarni — APPROVED** [review]
I agree with @rickyyx the limitation is a bit tricky, but I think it's fine for an experimental feature. In the future we can flesh out the use case more in the docs and provide a recommended workaround (I guess the workaround would involve working_dir or py_modules somehow?)

**Pull request merged**